src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs

```diff
@@ -177,21 +176,24 @@ public void Advance(int count)
         [MethodImpl(MethodImplOptions.NoInlining)]
         private void AdvanceToNextSpan(int count)
         {
-            while (!End && count > 0)
+            while (_moreData)
             {
-                if (CurrentSpanIndex < CurrentSpan.Length - count)
+                int remaining = CurrentSpan.Length - CurrentSpanIndex;
+
+                if (remaining > count)
                 {
                     CurrentSpanIndex += count;
                     count = 0;
                     break;
                 }
 
-                int remaining = (CurrentSpan.Length - CurrentSpanIndex);
-
-                CurrentSpanIndex += remaining;
                 count -= remaining;
+                Debug.Assert(count >= 0);
 
                 GetNextSpan();
+
+                if (count == 0)
```

Comment:
3 comparisons (if checks) instead of 4 for the common case (i.e. we advance to the next segment only once).

**Before:**
while (!End && count > 0) <= true and true
while (!End && count > 0) <= true and true
break

**After:**
while (_moreData) <= true
if (count == 0) <= false
while (_moreData) <= true
break